Add Block Storage API

gierens opened the pull request:
This starts adding the block storage API alias cinder. So far it implements rudimentary versions of `list_volumes`, `find_volumes` and `get_volume` with an equally minimal `Volume` struct. But I does compile and returns things.

Although the cinder API is very similar to nova's API, this is currently closer to the image module. For example, at the moment it goes directly to the `/volumes/detail` endpoint, rather than defining some kind of `VolumeSummary` for `/volumes` and offering a `detailed()` as the `compute` module does.

This is for simplicity, just to get things rolling. I need this at work, so I can invest a fair bit of time. I'll make it more consistent along the way.

dtantsur commented on src/block_storage/protocol.rs:
```diff
@@ -0,0 +1,84 @@
+
+use serde::Deserialize;
+
+// use super::super::common;
```
nit: remove?

gierens replied:
I removed the commented out import.

dtantsur commented on src/block_storage/protocol.rs:
```diff
@@ -0,0 +1,84 @@
+        Deleting = "deleting",
+        AwaitingTransfer = "awaiting-transfer",
+        Error = "error",
+        ErrorDeleting = "error_deleting",
```
The API constants actually use both underscores and dashes in different places?

gierens replied:
Yep, it does, see https://docs.openstack.org/api-ref/block-storage/v3/ and search for 'Volume statuses' ... I was surprised myself :D

dtantsur commented:
@gierens is this change still a draft? It looks pretty good to me. Could use a full acceptance test, and at least some unit test failures seem real (I need to check the others).

gierens commented:
> @gierens is this change still a draft? It looks pretty good to me.

Yeah, I was a little busy the last couple of weeks, that's why I didn't add anything more up until now. But I wanted to at least provide the most important volume functions before turning it into a normal PR. And then for snapshots and other stuff, this could go into a follow-up PR.

> Could use a full acceptance test, and at least some unit test failures seem real (I need to check the others).

I will now continue to work on this, I just added volume deletion and can also take a look at the failing jobs.

gierens commented:
I just added volume creation and with that I guess the most important functionality is present. With that I can now write a few integration tests and take a look at the still failing CI jobs. But after that this should be ready for review.

dtantsur commented:
The `No floating IP` issue has been haunting the project for a long time. It's possible that I'm simply too optimistic when assuming how long it takes to allocate such an IP. Chances are high, your patch is not at fault.

dtantsur commented:
Let's see if https://github.com/dtantsur/rust-openstack/pull/153 provides any improvement.

dtantsur commented:
@gierens if you rebase the patch, you'll pick up various CI fixes. Everything that fails afterwards is probably on you.

gierens commented:
@dtantsur thanks for the fix, that worked, great!

Then I would just add few integration tests for the block storage module and then mark the PR as ready for review later.

gierens commented:
Alright, I added the transparent property to all the remaining Volume fields, and two integration tests: The first creates a volume with just the required size argument, retrieves it, deletes and finally checks that its gone. And the other tests the creation with extra arguments like name and description.

dtantsur commented on src/block_storage/protocol.rs:
```diff
@@ -0,0 +1,201 @@
+
+//! JSON structures and protocol bits for the Block Storage API.
+
+#![allow(non_snake_case)]
```
What requires this?

gierens replied:
Probably just left-over when I initially copied code from another module, so I removed it.

dtantsur commented on src/block_storage/mod.rs:
```diff
@@ -0,0 +1,22 @@
+mod protocol;
+mod volumes;
+
+pub use self::protocol::{VolumeSortKey, VolumeStatus};
```
You may want to expose VolumeAttachment since it's used in the public API of Volume.

gierens replied:
done

dtantsur commented on src/block_storage/protocol.rs:
```diff
@@ -0,0 +1,201 @@
+
+/// A volume attachment.
+#[derive(Debug, Clone, Deserialize)]
+pub struct VolumeAttachment {
```
This is a part of the public API, does it make sense to make it as non-exhaustive to make sure we can add more fields without breaking anyone?

gierens replied:
What exactly do you mean by "non-exhaustive"?

I just added the fields that the cinder documentation show in an example, so would hope this is everything.

While checking the docs again I just realized, though, that the `host_name` should be an `Option`. Gonna fix that real quick.

gierens replied:
Ah now I know what you mean by non-exhaustive, saw it in the `protocol_enum` macro ... some of the Rust intricacies I still have to get familiar with :D ... I added `#[non_exhaustive]` to the struct.

dtantsur commented on src/block_storage/volumes.rs:
```diff
@@ -0,0 +1,453 @@
+
+    transparent_property! {
+        #[doc = "Volume links."]
+        links: ref Vec<protocol::Link>
```
So far, I've been avoiding exposing `links` as they're hard to work with in a type-safe fashion. Unless you have a strong case for using links, I'd omit this field.

gierens replied:
Done, I removed the `transparent_property` for both `links` and `volume_links`.

dtantsur commented on src/block_storage/volumes.rs:
```diff
@@ -0,0 +1,453 @@
+
+    transparent_property! {
+        #[doc = "Whether the volume is encrypted."]
+        encrypted: ref bool
```
You don't need to use ref with fields whose type is Copy, this leads to rather unergonomic `&bool` instead of just `bool`. The `ref` modifier is only needed for fields that cannot be just copied.

gierens replied:
done

dtantsur commented on src/block_storage/volumes.rs:
```diff
@@ -0,0 +1,453 @@
+
+    transparent_property! {
+        #[doc = "Size of the volume in GiB."]
+        size: ref u64
```
Same about ref

gierens replied:
done

dtantsur commented on src/block_storage/volumes.rs:
```diff
@@ -0,0 +1,453 @@
+
+    transparent_property! {
+        #[doc = "Status of the volume."]
+        status: ref protocol::VolumeStatus
```
Protocol enums are also Copy, you can omit ref here too.

gierens replied:
done

dtantsur commented on src/block_storage/volumes.rs:
```diff
@@ -0,0 +1,453 @@
+
+    transparent_property! {
+        #[doc = "Whether the volume is multi-attachable."]
+        multi_attachable: ref bool
```
Same

gierens replied:
done

dtantsur commented on src/block_storage/volumes.rs:
```diff
@@ -0,0 +1,453 @@
+
+    transparent_property! {
+        #[doc = "Whether the volume has shared targets."]
+        shared_targets: ref Option<bool>
```
Option<something Copy> is also Copy, so no ref again

gierens replied:
done

dtantsur commented on src/block_storage/volumes.rs:
```diff
@@ -0,0 +1,453 @@
+
+    transparent_property! {
+        #[doc = "Whether the volume consumes quota."]
+        consumes_quota: ref Option<bool>
```
Same here and below

gierens replied:
done

dtantsur commented on src/block_storage/volumes.rs:
```diff
@@ -0,0 +1,453 @@
+
+    transparent_property! {
+        #[doc = "When the volume was created."]
+        created_at: ref String
```
Would it be possible to convert this to DateTime automatically? You can check other resources for examples

gierens replied:
So, I thought about this and think I might have a solution. I created an enum `block_storage::protocol::DateTime` which wraps `chrono::DateTime<FixedOffset>` and `chrono::NaiveDateTime` in its variants `WithTz` and `WithoutTz` and implemented `Deserialize` accordingly. I also added other `impl`s that the `protocol_enum` would usually add.

This now works with our Zed cluster here where I don't seem to get timezone information, but would also work on clusters where you would.

dtantsur commented on src/block_storage/volumes.rs:
```diff
@@ -0,0 +1,453 @@
+
+    transparent_property! {
+        #[doc = "Whether the volume is bootable."]
+        bootable: ref String
```
Hmm, is it actually a string? From the name and description I'd expect a boolean...

gierens replied:
We could probably something similar to `Server.has_config_drive` and parse this into a `bool`. I wasn't sure here, because the cinder documentation is a little fuzzy.

The docs for the volume listing state: "Enables or disables the bootable attribute. You can boot an instance from a bootable volume." ... so this actually seems to be copy pasted from some other part of the docs.

In constract the nova docs on `config_drive` are pretty clear: "Indicates whether or not a config drive was used for this server. The value is True or an empty string. An empty string stands for False."

gierens replied:
I changed the `bootable` field to `bool` with the help of a new `bool_from_bootable_string` deserialization method.

dtantsur commented on src/block_storage/protocol.rs:
```diff
@@ -0,0 +1,201 @@
+    #[serde(rename = "os-vol-tenant-attr:tenant_id")]
+    pub tenant_id: Option<String>,
+    #[serde(rename = "os-vol-mig-status-attr:migstat")]
+    pub migstat: Option<String>, // consider enum
```
++ to a enum. We can add it later, of course, but that will be a breaking change (not the end of the world though).

In any case, could you rename it to migration_status or whatever fits? We're aiming for user-friendly names.

gierens replied:
Well, the reason I didn't use `migration_status` is because there already is a fields named `migration_status`. Any other suggestions? :D 

As for the enum, I will probably have to dig deeper then the docs here, because there is no table or list of valid entries.

dtantsur replied:
> Well, the reason I didn't use migration_status is because there already is a fields named migration_status. Any other suggestions? :D

Ehhhm, great :D What does this field actually do? If we don't know and cannot figure out, maybe skip it?

gierens replied:
Alright, I commented out this field and left a note that this is skipped due to the unclear use and uninformative naming.

dtantsur commented on src/block_storage/protocol.rs:
```diff
@@ -0,0 +1,201 @@
+    pub group_id: Option<String>,
+    pub volumes_links: Option<Vec<String>>,
+    pub provider_id: Option<String>,
+    pub service_uuid: Option<String>, // not optional in spec
```
Maybe rename to service_id? This is the only field that uses _uuid

gierens replied:
done

dtantsur commented:
Great job, thank you! I've left a few comments inline, could you check them?

Please let me know when I reach the point of abusing your patience, I can fix minor issues myself then :)

gierens commented:
@dtantsur Alright, I think I have addressed all comments and for the more complex cases also provided some info about how I solved them in my answers to the respective comments.

dtantsur commented on src/block_storage/protocol.rs:
```diff
@@ -0,0 +1,287 @@
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DateTime {
+    WithTz(chrono::DateTime<chrono::FixedOffset>),
+    WithoutTz(chrono::NaiveDateTime),
```
Won't this actually mean UTC?

dtantsur commented on src/block_storage/protocol.rs:
```diff
@@ -0,0 +1,287 @@
+    pub snapshot_id: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub backup_id: Option<String>,
+    pub name: String, // not optional in spec, but doesn't work with None/null, only with ""
```
Could be a helper for serialize_with (to convert None to "")

dtantsur commented:
@gierens I hope you don't mind me squashing the commits: 58 may easily be more than we had in rust-openstack before that :)

Let's continue the discussion around DateTime? I have a strong feeling that when they don't provide a timezone, they actually mean UTC. Could you try confirming?

I'll hold releasing the changes until we solve the DateTime problem and maybe figure out enums.

gierens commented:
@dtantsur That's totally fine. I tend to do atomic commits because squashing is a lot easier than taking commits apart. I probably could've done and interactive rebase in the end to cleanup the history a bit. But now that this one is in we can work on the rest in smaller chunks/PRs.

About the DateTime, yeah, I do share that suspicion. It's unfortunate the API reference is this vague. I'll experiment around a bit and get back to you then.

About releasing, sure totally understandable.
